internal/node/package_manager/subscriber.go

```diff
@@ -49,22 +49,58 @@ func (e *InstallAppExecutor) Execute(update *hdb.StateUpdate) error {
 	return nil
 }
 
-func (e *InstallAppExecutor) PostHook(update *hdb.StateUpdate) error {
+func (e *InstallAppExecutor) PostHook(update hdb.StateUpdateInternal) error {
 	var t node.StartInstallationTransition
-	err := json.Unmarshal(update.Transition, &t)
+	err := json.Unmarshal(update.Transition(), &t)
 	if err != nil {
 		return err
 	}
 
 	// After finishing the installation, update the application's lifecycle state
-	err = e.nodeController.FinishAppInstallation(t.UserID, t.AppInstallation.ID, t.RegistryURLBase, t.RegistryPackageID)
+	appInstallation := t.EnrichedData.AppState
+	err = e.nodeController.FinishAppInstallation(t.UserID, appInstallation.ID, appInstallation.RegistryURLBase, appInstallation.RegistryPackageID, t.StartAfterInstallation)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
+// FinishInstallExecutor is a state update executor that finishes the installation of an application.
+// Currently it is just run for it's PostHook, which will try to start the app if StartAfterInstallation is set.
```

Comment:
why can't we just make a PostHook on `InstallApp`?

Reply:
We don't want to run this unless the state for the app is already `installed` to maintain consistency. When the `PostHook` of `InstallAppExecutor` runs,[ it calls](https://github.com/eagraf/habitat-new/pull/23/files#diff-a5e50c6674dee457df8c8707dccd33e65ea6bfa9f2a38662b8247b927294111aR61) `nodeController.FinishAppInstallation`. With the way the controller works currently, it is making an aynchronous request to the state machine, so just because the function call completed successfully does not mean that the state was updated to `installed`.

We could create some sort of helper that makes state update requests to the state machine synchronous, but I think that would end up being more confusing than just creating another subscriber. The reason it would be more confusing is that Raft doesn't really provide a guarantee on how long the request will take to complete if there are multiple nodes involved, so we'd need to implement a timeout. I think timeouts should be avoided, because they can cause what appears to be inconsistent behavior to the client.

Reply:
Hm okay it just feels weird that we're running something just for its PostHook, like it seems in that case should make something that doesn't have a `PostHook` and just run it when we want. Especially because this is called `FinishInstall`, it should do the thing it says.


for e.g. can we have an executor that always runs on node start up and and starts any apps that need to be started?

Reply:
I think we can put this in the main execution for the subscriber if that makes it less confusing, there won't be any difference.